File: crates/dwall-settings/src/commands/solar.rs
//! Solar position commands

use dwall::{
    GeographicPositionProvider, SolarPosition, config::PositionSource, platform::Positioner,
};
use time::UtcDateTime;

use crate::error::DwallSettingsResult;

/// Sample interval for the diurnal sun path (10 minutes).
const PATH_SAMPLE_SECONDS: u64 = 10 * 60;
const DAY_SECONDS: u64 = 24 * 60 * 60;

#[tauri::command]
pub fn get_solar_position(
    position_source: PositionSource,
    timestamp: u64,
) -> DwallSettingsResult<SolarPosition> {
    let position_provider = GeographicPositionProvider::new(&position_source, Positioner::new());
    let position = position_provider.get_current_position()?;
    let date_time = UtcDateTime::from_timestamp(timestamp);
    Ok(SolarPosition::new(&position, &date_time))
}

/// Apparent solar positions sampled over a full day, centred on `timestamp`.
///
/// Used to draw the diurnal sun-trajectory preview: unlike a theme's
/// `solar.json` (which holds the wallpaper keyframes), this is the real path
/// for the observer's location and date, so it always forms a smooth ellipse.
#[tauri::command]
pub fn get_solar_path(
    position_source: PositionSource,
    timestamp: u64,
) -> DwallSettingsResult<Vec<SolarPosition>> {
    let position_provider = GeographicPositionProvider::new(&position_source, Positioner::new());
    let position = position_provider.get_current_position()?;

    let start = timestamp.saturating_sub(DAY_SECONDS / 2);
    let samples = (0..=DAY_SECONDS / PATH_SAMPLE_SECONDS)
        .map(|i| {
            let date_time = UtcDateTime::from_timestamp(start + i * PATH_SAMPLE_SECONDS);
            SolarPosition::new(&position, &date_time)
        })
        .collect();

    Ok(samples)
}

File: src/ipc/solar.ts
import { invoke } from "@tauri-apps/api/core";

import type { PositionSource, SolarPosition } from "./types";

export const getSolarPosition = async (
  positionSource: PositionSource,
  timestamp: number,
): Promise<SolarPosition> =>
  invoke("get_solar_position", { positionSource, timestamp });

export const currentSolarPosition = async (
  positionSource: PositionSource,
): Promise<SolarPosition> =>
  getSolarPosition(positionSource, Math.floor(Date.now() / 1000));

/** Apparent solar positions sampled over a full day, centred on `timestamp`. */
export const getSolarPath = async (
  positionSource: PositionSource,
  timestamp: number,
): Promise<SolarPosition[]> =>
  invoke("get_solar_path", { positionSource, timestamp });
